File: Nudge-Neurodivergent-task-tracker/TaskBreakdown.java
import java.util.List;
import java.util.ArrayList;

/**
 * Helps break down large tasks into manageable subtasks for executive function support
 */
public class TaskBreakdown {
    public enum Priority {
        URGENT_IMPORTANT("🔴 Urgent & Important", 1),
        IMPORTANT_NOT_URGENT("🟡 Important, Not Urgent", 2),
        URGENT_NOT_IMPORTANT("🟠 Urgent, Not Important", 3),
        NEITHER("🟢 Neither Urgent nor Important", 4);
        
        private final String displayName;
        private final int value;
        
        Priority(String displayName, int value) {
            this.displayName = displayName;
            this.value = value;
        }
        
        public String getDisplayName() { return displayName; }
        public int getValue() { return value; }
    }
    
    public enum Difficulty {
        VERY_EASY("🟢 Very Easy - 5 min or less"),
        EASY("🔵 Easy - 15 min"),
        MODERATE("🟡 Moderate - 30-45 min"),
        HARD("🟠 Hard - 1-2 hours"),
        VERY_HARD("🔴 Very Hard - Multiple sessions");
        
        private final String displayName;
        
        Difficulty(String displayName) {
            this.displayName = displayName;
        }
        
        public String getDisplayName() { return displayName; }
    }
    
    public static class Subtask {
        private String description;
        private Difficulty difficulty;
        private int estimatedMinutes;
        private boolean isCompleted;
        private String context; // What tools/materials needed
        
        public Subtask(String description, Difficulty difficulty, int estimatedMinutes, String context) {
            this.description = description;
            this.difficulty = difficulty;
            this.estimatedMinutes = estimatedMinutes;
            this.context = context;
            this.isCompleted = false;
        }
        
        public void markCompleted() { this.isCompleted = true; }
        
        // Getters
        public String getDescription() { return description; }
        public Difficulty getDifficulty() { return difficulty; }
        public int getEstimatedMinutes() { return estimatedMinutes; }
        public boolean isCompleted() { return isCompleted; }
        public String getContext() { return context; }
        
        @Override
        public String toString() {
            String status = isCompleted ? "✅" : "⬜";
            return String.format("%s %s (%s, ~%d min) - %s", 
                status, description, difficulty.getDisplayName(), estimatedMinutes, context);
        }
    }
    
    public static class TaskPlan {
        private Task originalTask;
        private Priority priority;
        private Difficulty overallDifficulty;
        private List<Subtask> subtasks;
        private String preparationNotes;
        private String contextSwitchNotes;
        
        public TaskPlan(Task originalTask) {
            this.originalTask = originalTask;
            this.subtasks = new ArrayList<>();
            this.priority = Priority.IMPORTANT_NOT_URGENT;
            this.overallDifficulty = Difficulty.MODERATE;
        }
        
        public void addSubtask(String description, Difficulty difficulty, int estimatedMinutes, String context) {
            subtasks.add(new Subtask(description, difficulty, estimatedMinutes, context));
        }
        
        public void removeSubtask(int index) {
            if (index >= 0 && index < subtasks.size()) {
                subtasks.remove(index);
            }
        }
        
        public void completeSubtask(int index) {
            if (index >= 0 && index < subtasks.size()) {
                subtasks.get(index).markCompleted();
            }
        }
        
        public int getCompletionPercentage() {
            if (subtasks.isEmpty()) return 0;
            
            long completed = subtasks.stream().mapToLong(subtask -> subtask.isCompleted() ? 1 : 0).sum();
            return (int) ((completed * 100) / subtasks.size());
        }
        
        public int getTotalEstimatedTime() {
            return subtasks.stream().mapToInt(Subtask::getEstimatedMinutes).sum();
        }
        
        public List<Subtask> getNextActions(int maxItems) {
            return subtasks.stream()
                .filter(subtask -> !subtask.isCompleted())
                .limit(maxItems)
                .toList();
        }
        
        public boolean isFullyCompleted() {
            return !subtasks.isEmpty() && subtasks.stream().allMatch(Subtask::isCompleted);
        }
        
        public void displayPlan() {
            System.out.println("\n📋 TASK BREAKDOWN: " + originalTask.getName());
            System.out.println("🎯 Priority: " + priority.getDisplayName());
            System.out.println("📊 Difficulty: " + overallDifficulty.getDisplayName());
            System.out.println("⏱️ Total estimated time: " + getTotalEstimatedTime() + " minutes");
            System.out.println("📈 Progress: " + getCompletionPercentage() + "%");
            
            if (preparationNotes != null && !preparationNotes.isEmpty()) {
                System.out.println("\n🎒 Preparation needed:");
                System.out.println("   " + preparationNotes);
            }
            
            if (contextSwitchNotes != null && !contextSwitchNotes.isEmpty()) {
                System.out.println("\n🔄 Context switching notes:");
                System.out.println("   " + contextSwitchNotes);
            }
            
            System.out.println("\n📝 Subtasks:");
            System.out.println("-".repeat(50));
            for (int i = 0; i < subtasks.size(); i++) {
                System.out.printf("%d. %s\n", i + 1, subtasks.get(i));
            }
            
            List<Subtask> nextActions = getNextActions(3);
            if (!nextActions.isEmpty()) {
                System.out.println("\n⭐ Next 3 actions to focus on:");
                for (int i = 0; i < nextActions.size(); i++) {
                    System.out.printf("   %d. %s\n", i + 1, nextActions.get(i).getDescription());
                }
            }
        }
        
        // Getters and setters
        public Task getOriginalTask() { return originalTask; }
        public Priority getPriority() { return priority; }
        public Difficulty getOverallDifficulty() { return overallDifficulty; }
        public List<Subtask> getSubtasks() { return new ArrayList<>(subtasks); }
        public String getPreparationNotes() { return preparationNotes; }
        public String getContextSwitchNotes() { return contextSwitchNotes; }
        
        public void setPriority(Priority priority) { this.priority = priority; }
        public void setOverallDifficulty(Difficulty difficulty) { this.overallDifficulty = difficulty; }
        public void setPreparationNotes(String notes) { this.preparationNotes = notes; }
        public void setContextSwitchNotes(String notes) { this.contextSwitchNotes = notes; }
    }
    
    private List<TaskPlan> taskPlans;
    
    public TaskBreakdown() {
        this.taskPlans = new ArrayList<>();
    }
    
    public TaskPlan createTaskPlan(Task task) {
        TaskPlan plan = new TaskPlan(task);
        taskPlans.add(plan);
        return plan;
    }
    
    public void removeTaskPlan(Task task) {
        taskPlans.removeIf(plan -> plan.getOriginalTask().getId() == task.getId());
    }
    
    public TaskPlan findTaskPlan(Task task) {
        return taskPlans.stream()
            .filter(plan -> plan.getOriginalTask().getId() == task.getId())
            .findFirst()
            .orElse(null);
    }
    
    public List<TaskPlan> getAllTaskPlans() {
        return new ArrayList<>(taskPlans);
    }
    
    public List<TaskPlan> getTaskPlansByPriority(Priority priority) {
        return taskPlans.stream()
            .filter(plan -> plan.getPriority() == priority)
            .toList();
    }
    
    public TaskPlan suggestBreakdown(Task task) {
        TaskPlan plan = createTaskPlan(task);
        
        // AI-assisted task breakdown based on task name and category
        String taskName = task.getName().toLowerCase();
        TaskCategory category = task.getCategory();
        
        if (taskName.contains("clean") || taskName.contains("organize")) {
            plan.addSubtask("Gather cleaning supplies", Difficulty.VERY_EASY, 5, "Cleaning supplies");
            plan.addSubtask("Clear surfaces", Difficulty.EASY, 15, "Boxes/containers for sorting");
            plan.addSubtask("Deep clean", Difficulty.MODERATE, 30, "Cleaning products");
            plan.addSubtask("Organize and put away", Difficulty.EASY, 20, "Storage solutions");
            plan.setPreparationNotes("Have cleaning supplies ready, put on energizing music");
            
        } else if (taskName.contains("call") || taskName.contains("phone")) {
            plan.addSubtask("Find contact information", Difficulty.VERY_EASY, 5, "Phone/contacts app");
            plan.addSubtask("Write down key points to discuss", Difficulty.EASY, 10, "Notepad/app");
            plan.addSubtask("Make the call", Difficulty.MODERATE, 15, "Quiet space, charged phone");
            plan.addSubtask("Follow up notes", Difficulty.VERY_EASY, 5, "Calendar/reminder app");
            plan.setContextSwitchNotes("Schedule for when you have good energy and quiet space");
            
        } else if (taskName.contains("email") || taskName.contains("write")) {
            plan.addSubtask("Open email/document", Difficulty.VERY_EASY, 2, "Computer/phone");
            plan.addSubtask("Draft main points", Difficulty.EASY, 15, "Notes or outline");
            plan.addSubtask("Write full content", Difficulty.MODERATE, 25, "Focused environment");
            plan.addSubtask("Review and edit", Difficulty.EASY, 10, "Fresh eyes");
            plan.addSubtask("Send/submit", Difficulty.VERY_EASY, 3, "Double-check recipients");
            
        } else if (taskName.contains("appointment") || taskName.contains("schedule")) {
            plan.addSubtask("Check available times", Difficulty.VERY_EASY, 5, "Calendar app");
            plan.addSubtask("Find contact information", Difficulty.VERY_EASY, 5, "Phone/website");
            plan.addSubtask("Make the call/booking", Difficulty.EASY, 10, "Phone/computer");
            plan.addSubtask("Add to calendar", Difficulty.VERY_EASY, 3, "Calendar app");
            plan.addSubtask("Set reminders", Difficulty.VERY_EASY, 2, "Phone/calendar");
            
        } else {
            // Generic breakdown
            plan.addSubtask("Gather materials/information", Difficulty.EASY, 10, "Whatever task requires");
            plan.addSubtask("Start main work", Difficulty.MODERATE, 30, "Focused environment");
            plan.addSubtask("Review and finalize", Difficulty.EASY, 15, "Fresh perspective");
            plan.setPreparationNotes("Make sure you have everything needed before starting");
        }
        
        // Set priority based on task deadline and category
        if (task.isOverdue()) {
            plan.setPriority(Priority.URGENT_IMPORTANT);
        } else if (task.isDueSoon()) {
            plan.setPriority(Priority.URGENT_NOT_IMPORTANT);
        } else if (category == TaskCategory.MEDICATION || category == TaskCategory.HEALTH) {
            plan.setPriority(Priority.IMPORTANT_NOT_URGENT);
        } else {
            plan.setPriority(Priority.NEITHER);
        }
        
        return plan;
    }
    
    public void displayPriorityMatrix() {
        System.out.println("\n📊 EISENHOWER PRIORITY MATRIX");
        System.out.println("=".repeat(60));
        
        for (Priority priority : Priority.values()) {
            List<TaskPlan> plans = getTaskPlansByPriority(priority);
            if (!plans.isEmpty()) {
                System.out.println("\n" + priority.getDisplayName() + ":");
                for (TaskPlan plan : plans) {
                    System.out.printf("   📋 %s (%d%% complete)\n", 
                        plan.getOriginalTask().getName(), 
                        plan.getCompletionPercentage());
                }
            }
        }
        
        System.out.println("\n💡 FOCUS RECOMMENDATIONS:");
        System.out.println("   🔴 Do urgent & important tasks first");
        System.out.println("   🟡 Schedule time for important, non-urgent tasks");
        System.out.println("   🟠 Delegate or minimize urgent, unimportant tasks");
        System.out.println("   🟢 Eliminate or do in spare time");
    }
    
    public List<String> getExecutiveFunctionTips(EnergyTracker.EnergyLevel energy) {
        List<String> tips = new ArrayList<>();
        
        switch (energy) {
            case VERY_LOW:
                tips.add("🌸 Pick the tiniest possible next step");
                tips.add("📱 Use voice memos instead of writing");
                tips.add("🎵 Try body doubling or background sounds");
                break;
            case LOW:
                tips.add("⏰ Set a 10-minute timer and just start");
                tips.add("📝 Write down the very next physical action needed");
                tips.add("🔄 Use the 2-minute rule - if it takes less than 2 min, do it now");
                break;
            case MODERATE:
                tips.add("📋 Focus on one subtask at a time");
                tips.add("🍅 Try a Pomodoro session");
                tips.add("🎯 Set mini-goals within larger tasks");
                break;
            case GOOD:
            case HIGH:
                tips.add("🚀 Great time for tackling complex multi-step tasks");
                tips.add("📈 Consider batch processing similar subtasks");
                tips.add("🎪 Use this energy to prepare future low-energy days");
                break;
        }
        
        tips.add("🧠 Remember: your brain works differently, and that's okay");
        tips.add("💪 Progress over perfection, always");
        
        return tips;
    }
}

File: Nudge-Neurodivergent-task-tracker/task_breakdown.py
"""
Task breakdown and executive function support
"""

from datetime import datetime
from enum import Enum
from typing import List, Optional, Dict
from dataclasses import dataclass
import random
from energy_tracker import EnergyLevel
from task_manager import Task


class Difficulty(Enum):
    """Task difficulty levels"""
    EASY = ("🟢 Easy", "Low mental load")
    MODERATE = ("🟡 Moderate", "Medium effort required")
    HARD = ("🔴 Hard", "High concentration needed")
    
    def __init__(self, display: str, description: str):
        self.display = display
        self.description = description


class Priority(Enum):
    """Priority levels using Eisenhower matrix"""
    URGENT_IMPORTANT = ("🔥 Do First", "Urgent & Important")
    NOT_URGENT_IMPORTANT = ("📅 Schedule", "Important, Not Urgent")
    URGENT_NOT_IMPORTANT = ("🤝 Delegate", "Urgent, Not Important")
    NOT_URGENT_NOT_IMPORTANT = ("🗑️ Eliminate", "Neither Urgent nor Important")
    
    def __init__(self, action: str, description: str):
        self.action = action
        self.description = description


@dataclass
class Subtask:
    """A subtask within a larger task breakdown"""
    description: str
    difficulty: Difficulty
    estimated_minutes: int
    context_notes: str
    completed: bool = False
    completed_at: Optional[datetime] = None
    
    def mark_completed(self):
        """Mark subtask as completed"""
        self.completed = True
        self.completed_at = datetime.now()


@dataclass
class TaskPlan:
    """A complete task breakdown plan"""
    original_task: Task
    subtasks: List[Subtask]
    priority: Priority
    context_switching_notes: str
    created_at: datetime
    
    def display_plan(self):
        """Display the task breakdown plan"""
        print(f"\n📝 TASK BREAKDOWN: {self.original_task.name}")
        print("=" * 50)
        print(f"🎯 Priority: {self.priority.action} - {self.priority.description}")
        print(f"⏱️ Total estimated time: {sum(s.estimated_minutes for s in self.subtasks)} minutes")
        print(f"📋 Progress: {self.get_completion_percentage():.0f}% complete")
        
        if self.context_switching_notes:
            print(f"\n🧠 Context Notes: {self.context_switching_notes}")
        
        print("\n📋 SUBTASKS:")
        for i, subtask in enumerate(self.subtasks, 1):
            status = "✅" if subtask.completed else "⚪"
            print(f"  {status} {i}. {subtask.description}")
            print(f"      {subtask.difficulty.display} | {subtask.estimated_minutes}min")
            if subtask.context_notes:
                print(f"      💡 {subtask.context_notes}")
    
    def get_completion_percentage(self) -> float:
        """Get completion percentage"""
        if not self.subtasks:
            return 0
        completed = sum(1 for s in self.subtasks if s.completed)
        return (completed / len(self.subtasks)) * 100
    
    def is_fully_completed(self) -> bool:
        """Check if all subtasks are completed"""
        return all(s.completed for s in self.subtasks)
    
    def complete_subtask(self, index: int) -> bool:
        """Complete a subtask by index"""
        if 0 <= index < len(self.subtasks):
            self.subtasks[index].mark_completed()
            return True
        return False
    
    def add_subtask(self, description: str, difficulty: Difficulty, 
                   estimated_minutes: int, context_notes: str = ""):
        """Add a new subtask"""
        subtask = Subtask(description, difficulty, estimated_minutes, context_notes)
        self.subtasks.append(subtask)


class TaskBreakdown:
    """Provides AI-powered task breakdown and executive function support"""
    
    def __init__(self):
        self.task_plans: Dict[str, TaskPlan] = {}  # task_id -> TaskPlan
    
    def suggest_breakdown(self, task: Task) -> TaskPlan:
        """Create an AI-powered task breakdown"""
        # Generate breakdown based on task type and complexity
        subtasks = self._generate_subtasks(task)
        priority = self._determine_priority(task)
        context_notes = self._generate_context_notes(task)
        
        plan = TaskPlan(
            original_task=task,
            subtasks=subtasks,
            priority=priority,
            context_switching_notes=context_notes,
            created_at=datetime.now()
        )
        
        self.task_plans[task.task_id] = plan
        return plan
    
    def _generate_subtasks(self, task: Task) -> List[Subtask]:
        """Generate subtasks based on task characteristics"""
        task_name = task.name.lower()
        subtasks = []
        
        # General task breakdown patterns
        if any(word in task_name for word in ['email', 'message', 'contact', 'call']):
            subtasks = [
                Subtask("Gather contact information", Difficulty.EASY, 3, "Check phone/email"),
                Subtask("Draft message outline", Difficulty.MODERATE, 10, "Key points to cover"),
                Subtask("Write/make contact", Difficulty.MODERATE, 15, "Be clear and direct"),
                Subtask("Follow up if needed", Difficulty.EASY, 5, "Set reminder for response")
            ]
        elif any(word in task_name for word in ['clean', 'organize', 'tidy']):
            subtasks = [
                Subtask("Gather supplies", Difficulty.EASY, 5, "Cleaning materials, storage"),
                Subtask("Set up music/podcast", Difficulty.EASY, 2, "Make it enjoyable"),
                Subtask("Start with smallest area", Difficulty.MODERATE, 15, "Build momentum"),
                Subtask("Sort and decide", Difficulty.HARD, 20, "Keep/donate/trash decisions"),
                Subtask("Put everything away", Difficulty.MODERATE, 10, "Everything has a place")
            ]
        elif any(word in task_name for word in ['research', 'study', 'learn']):
            subtasks = [
                Subtask("Define specific questions", Difficulty.MODERATE, 10, "What exactly do you need?"),
                Subtask("Find reliable sources", Difficulty.MODERATE, 15, "Quality over quantity"),
                Subtask("Take notes actively", Difficulty.HARD, 30, "Summarize key points"),
                Subtask("Organize findings", Difficulty.MODERATE, 10, "Make it usable later")
            ]
        elif any(word in task_name for word in ['write', 'document', 'report']):
            subtasks = [
                Subtask("Brain dump ideas", Difficulty.EASY, 10, "Don't edit, just write"),
                Subtask("Create outline", Difficulty.MODERATE, 15, "Structure your thoughts"),
                Subtask("Write first draft", Difficulty.HARD, 45, "Progress over perfection"),
                Subtask("Review and edit", Difficulty.MODERATE, 20, "Read aloud for flow"),
                Subtask("Final proofread", Difficulty.EASY, 10, "Fresh eyes if possible")
            ]
        else:
            # Generic breakdown
            estimated_total = task.estimated_duration_minutes
            subtasks = [
                Subtask("Prepare and gather materials", Difficulty.EASY, 
                       max(5, estimated_total // 6), "Set up workspace"),
                Subtask("Start the main work", Difficulty.MODERATE,
                       max(15, estimated_total // 2), "Focus on core task"),
                Subtask("Complete and wrap up", Difficulty.MODERATE,
                       max(10, estimated_total // 3), "Finish and clean up")
            ]
        
        return subtasks
    
    def _determine_priority(self, task: Task) -> Priority:
        """Determine task priority using Eisenhower matrix"""
        # Simple heuristic based on deadline and category
        is_urgent = task.is_due_soon(hours=48)  # Due within 2 days
        
        # Important categories
        important_categories = [
            'MEDICATION', 'HEALTH', 'WORK'
        ]
        is_important = task.category.name in important_categories
        
        if is_urgent and is_important:
            return Priority.URGENT_IMPORTANT
        elif not is_urgent and is_important:
            return Priority.NOT_URGENT_IMPORTANT
        elif is_urgent and not is_important:
            return Priority.URGENT_NOT_IMPORTANT
        else:
            return Priority.NOT_URGENT_NOT_IMPORTANT
    
    def _generate_context_notes(self, task: Task) -> str:
        """Generate context switching and preparation notes"""
        notes = []
        
        # Time-based notes
        if task.estimated_duration_minutes > 60:
            notes.append("🍅 Consider using Pomodoro technique for focus")
        
        # Category-specific notes
        category_notes = {
            'WORK': "📧 Close personal tabs, use work environment",
            'CREATIVE': "🎨 Gather inspiration, prepare creative space",
            'SOCIAL': "👥 Check your social energy levels first",
            'CHORES': "🎵 Put on energizing music or podcast",
            'HEALTH': "🏥 Gather any needed documents or information"
        }
        
        if task.category.name in category_notes:
            notes.append(category_notes[task.category.name])
        
        # General preparation
        notes.extend([
            "📱 Put phone on focus mode",
            "💧 Have water nearby",
            "⏰ Set realistic time expectations"
        ])
        
        return " | ".join(notes)
    
    def get_executive_function_tips(self, energy_level: EnergyLevel) -> List[str]:
        """Get executive function tips based on current energy"""
        base_tips = [
            "🧠 Break tasks into 2-minute micro-steps",
            "⏰ Use external timers for time awareness",
            "📝 Write down every step (external memory)",
            "🔄 Change environment when stuck",
            "🎯 Do the smallest possible next action"
        ]
        
        if energy_level.level_value <= 2:  # Low energy
            base_tips.extend([
                "🌸 Pick the tiniest possible next step",
                "📱 Use voice memos instead of writing",
                "🎵 Try body doubling or background sounds",
                "🧠 Remember: your brain works differently, and that's okay",
                "💪 Progress over perfection, always"
            ])
        elif energy_level.level_value >= 4:  # High energy
            base_tips.extend([
                "⚡ Batch similar tasks together",
                "🚀 Tackle avoided or complex tasks",
                "📋 Use this energy for planning sessions",
                "⚠️ Set boundaries to prevent burnout"
            ])
        
        return base_tips
    
    def find_task_plan(self, task: Task) -> Optional[TaskPlan]:
        """Find existing task plan"""
        return self.task_plans.get(task.task_id)
    
    def remove_task_plan(self, task: Task):
        """Remove task plan when task is deleted"""
        if task.task_id in self.task_plans:
            del self.task_plans[task.task_id]
    
    def get_all_active_plans(self) -> List[TaskPlan]:
        """Get all active task plans"""
        return list(self.task_plans.values())